spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,487 @@
+    )
+
+    
+    // either timeout height or timestamp must be non-zero
+    // if the upgrade feature is implemented on the TRY chain, then a relayer may submit a TRY transaction after the timeout.
+    // this will restore the connection on the executing chain and allow counterparty to use the CancelUpgradeMsg to restore their connection.
+    restoreConnectionUnless(timeoutHeight != 0 || timeoutTimestamp != 0)
+    upgradeTimeout = UpgradeTimeout{
+        timeoutHeight: timeoutHeight,
+        timeoutTimestamp: timeoutTimestamp,
+    }
+
+    // verify that counterparty connection unmodifiable fields have not changed and counterparty state
+    // is UPGRADE_INIT
+    restoreConnectionUnless(
+        counterpartyConnection.state == UPGRADE_INIT &&
+        counterpartyConnection.counterpartyConnectionIdentifier == identifier &&
+        counterpartyConnection.clientIdentifier == currentConnection.counterpartyClientIdentifier &&
+        counterpartyConnection.counterpartyClientIdentifier == currentConnection.clientIdentifier
+    )
+
+    // counterparty-specified timeout must not have exceeded
+    restoreConnectionUnless(
+        timeoutHeight > getCurrentHeight() ||
+        timeoutTimestamp > currentTimestamp()
+    )
+
+    // verify chosen versions are compatible
+    versionsIntersection = intersection(counterpartyConnection.version, proposedUpgradeConnection.version)
+    version = pickVersion(versionsIntersection) // aborts transaction if there is no intersection
+
+    // both connection ends must be mutually compatible.
+    // this function has been left unspecified since it will depend on the specific structure of the new connection.
+    // It is the responsibility of implementations to make sure that verification that the proposed new connections
+    // on either side are correctly constructed according to the new version selected.
+    restoreConnectionUnless(IsCompatible(counterpartyConnection, proposedUpgradeConnection))
+
+    // verify proofs of counterparty state
+    abortTransactionUnless(verifyConnectionState(currentConnection, proofHeight, proofConnection, currentConnection.counterpartyConnectionIdentifier, proposedUpgradeConnection))
+    abortTransactionUnless(verifyConnectionUpgradeTimeout(currentConnection, proofHeight, proofUpgradeTimeout,  upgradeTimeout))
+ 
+    provableStore.set(connectionPath(identifier), proposedUpgradeConnection)
+}
+```
+
+NOTE: It is up to individual implementations how they will provide access-control to the `ConnUpgradeTry` function. E.g. chain governance, permissioned actor, DAO, etc. A chain may decide to have permissioned **or** permissionless `UpgradeTry`. In the permissioned case, both chains must explicitly consent to the upgrade, in the permissionless case; one chain initiates the upgrade and the other chain agrees to the upgrade by default. In the permissionless case, a relayer may submit the `ConnUpgradeTry` datagram.
+
+
+```typescript
+function connUpgradeAck(
```

Comment:
so the timeout value only applies to the first two steps of the handshake?

Reply:
Yes because you need liveness to be restored for timeouts to be processed, and if liveness is restored you may as well just continue with the handshake anyway.

The timeout is really only useful in this case, when it is possible that we stall indefinitely which is only possible on TRY step

Reply:
See Timeout Upgrade Process